transcriptic/sampledata/_data/p123-runs.json

```diff
@@ -1 +1 @@
r_type":"User","progress":0,"project_id":"p123","properties":{},"protocol_id":null,"quote":{"service_level":"HIGHLEVEL"},"request":"{\"type\":\"custom\",\"details\":\"My project...\",\"attachments\":null}","request_type":"custom","results":{},"scheduled_to_start_at":null,"scheduled_workcell":null,"sent_to_workcell_by_id":null,"sent_to_workcell_by_user_id":null,"started_at":null,"status":"canceled","success":null,"success_notes":null,"test_mode":false,"title":"Sample Run","updated_at":"2017-02-13T12:35:23.612-08:00","priority":"Medium","requested_at":null,"reject_reason":null,"reject_description":null,"lab_id":"lb1fej8nubcf3k3","protocol_name":null,"contextual_custom_properties_configs":[],"contextual_custom_properties":[],"billing_valid?":true,"pending_shipment_ids":[],"unrealized_input_containers_count":0,"organization_name":"Aperture Science","organization_id":"org1"}}],"meta":{"record_count":1},"links":{"first":"https://webapp.staging.strateos.com/api/runs?fields%5Bruns%5D=aborted_at%2Caborted_reason%2Caccepted_at%2Caccepted_by_id%2Cattachments%2Cbsl%2Ccanceled_at%2Ccanceled_reason%2Ccommitted_at%2Ccompleted_at%2Cconversation_id%2Ccreated_at%2Cdraft_quote%2Cestimated_run_time_cache%2Cexecution%2Cflagged%2Cfriendly_status%2Cinternal_run%2Claunch_request_id%2Cout_section%2Cowner_id%2Cowner_type%2Cprogress%2Cproject_id%2Cproperties%2Cprotocol_id%2Cquote%2Crequest%2Crequest_type%2Cresults%2Cscheduled_to_start_at%2Cscheduled_workcell%2Csent_to_workcell_by_id%2Csent_to_workcell_by_user_id%2Cstarted_at%2Cstatus%2Csuccess%2Csuccess_notes%2Ctest_mode%2Ctitle%2Cupdated_at%2Cpriority%2Crequested_at%2Creject_reason%2Creject_description%2Clab_id%2Csuccessors_deep%2Cprotocol_name%2Ccontextual_custom_properties_configs%2Ccontextual_custom_properties%2Cbilling_valid%3F%2Cpending_shipment_ids%2Cunrealized_input_containers_count%2Corganization_name%2Corganization_id&filter%5Bproject_id%5D=p16rubzndwp4q&page%5Blimit%5D=12&page%5Boffset%5D=0","last":"https://webapp.staging.strateos.com/api/runs?fields%5Bruns%5D=aborted_at%2Caborted_reason%2Caccepted_at%2Caccepted_by_id%2Cattachments%2Cbsl%2Ccanceled_at%2Ccanceled_reason%2Ccommitted_at%2Ccompleted_at%2Cconversation_id%2Ccreated_at%2Cdraft_quote%2Cestimated_run_time_cache%2Cexecution%2Cflagged%2Cfriendly_status%2Cinternal_run%2Claunch_request_id%2Cout_section%2Cowner_id%2Cowner_type%2Cprogress%2Cproject_id%2Cproperties%2Cprotocol_id%2Cquote%2Crequest%2Crequest_type%2Cresults%2Cscheduled_to_start_at%2Cscheduled_workcell%2Csent_to_workcell_by_id%2Csent_to_workcell_by_user_id%2Cstarted_at%2Cstatus%2Csuccess%2Csuccess_notes%2Ctest_mode%2Ctitle%2Cupdated_at%2Cpriority%2Crequested_at%2Creject_reason%2Creject_description%2Clab_id%2Csuccessors_deep%2Cprotocol_name%2Ccontextual_custom_properties_configs%2Ccontextual_custom_properties%2Cbilling_valid%3F%2Cpending_shipment_ids%2Cunrealized_input_containers_count%2Corganization_name%2Corganization_id&filter%5Bproject_id%5D=p16rubzndwp4q&page%5Blimit%5D=12&page%5Boffset%5D=0"}}
```

Comment:
Is this an accurate response from `api/runs?filter[project_id]={project_id}&fields[runs]=id,title,status,created_at,completed_at`? 
1. There should be some fields filtered out, right?
2. Why not make this an array and not convert it to an array in the test?

Reply:
Are we suppose to maintain other fields? Because then, every change made to the response, like an attribute removal, would be a breaking change.

Reply:
But it's parsed initially here:

```
    def runs(self, project_id=None):
        """Get list of runs in project"""
        route = self.get_route("get_project_runs", project_id=project_id)
        err_default = (
            f"There was an error fetching the runs in project " f"{project_id}"
        )
        results = []
        while True:
            response = self.get(
                route,
                status_response={
                    "200": lambda resp: resp.json(),
                    "default": lambda resp: RuntimeError(err_default),
                },
            )
            results.extend(
                [{**data["attributes"], "id": data["id"]} for data in response["data"]]
            )
            if "next" in response["links"]:
                route = response["links"]["next"]
            else:
                break
        return results
```

Reply:
The response from the JSONAPI endpoint is different than the non-JSONAPI endpoint. The parsing above is, again, only parsing a subset of the response.

What I'm saying is, let's say there is a user that is using the current `runs` function to retrieve all runs and then has some logic that parses the `refs`, for example. `refs` are included in the old response, but not in the new. Same goes for `instructions` and some other attributes/nested-objects.

Old (apologies for this mega-paste):
```
[
    {
        "id": "r1gvxaxcjwqh3r",
        "status": "canceled",
        "title": "Synth_Luke-Test-002",
        "created_at": "2022-04-04T15:25:24.666-07:00",
        "updated_at": "2022-04-04T15:30:20.870-07:00",
        "completed_at": null,
        "conversation_id": "conv1gvxaxcn9d6dt",
        "quote": {
            "items": [
                {
                    "cost": "1.75",
                    "quantity": 1,
                    "title": "Workcell Time",
                    "run_id": "r1gvxaxcjwqh3r",
                    "run_credit_applicable": true
                },
                {
                    "cost": "744.46",
                    "quantity": 1,
                    "title": "Reagents & Consumables",
                    "run_id": "r1gvxaxcjwqh3r",
                    "run_credit_applicable": false
                }
            ],
            "ppp": false
        },
        "results": {},
        "test_mode": false,
        "accepted_by_id": "u1fkpc5skavpkm",
        "accepted_at": "2022-04-04T15:25:24.664-07:00",
        "started_at": null,
        "canceled_at": "2022-04-04T15:30:20.856-07:00",
        "aborted_at": null,
        "predecessor_id": null,
        "draft_quote": null,
        "progress": 0,
        "protocol_id": "pr1grd45pr4xuck",
        "request_type": "protocol",
        "launch_request_id": "lr1gvxax9rjsfyc",
        "flagged": false,
        "estimated_run_time_cache": null,
        "scheduled_to_start_at": null,
        "properties": {},
        "internal_run": true,
        "bsl": 1,
        "success": null,
        "success_notes": null,
        "priority": "Medium",
        "requested_at": null,
        "total_cost": "746.21",
        "has_quote?": true,
        "can_cancel?": false,
        "warnings": [],
        "pending_shipment_ids": [],
        "billing_valid?": true,
        "unrealized_input_containers": [],
        "dependents": [],
        "datasets": [],
        "refs": [
            {
                "name": "Luke-Test-002-crude-LCMS-1",
                "container_id": null,
                "new_container_type": "hplc-vial",
                "destiny": {
                    "store": {
                        "where": "quarantine",
                        "shaking": false
                    }
                },
                "container_type": {
                    "id": "hplc-vial",
                    "name": "HPLC Vial",
                    "well_count": 1,
                    "well_depth_mm": "31.4",
                    "well_volume_ul": "2000.0",
                    "capabilities": [
                        "agitate",
                        "cover",
                        "dispense",
                        "evaporate",
                        "image",
                        "incubate",
                        "lcms",
                        "liquid_handle",
                        "measure_mass",
                        "provision",
                        "spe",
                        "spin",
                        "uncover"
                    ],
                    "shortname": "hplc-vial",
                    "col_count": 1,
                    "is_tube": true,
                    "acceptable_lids": [
                        "screw-cap"
                    ],
                    "height_mm": "33.73",
                    "vendor": "unknown",
                    "catalog_number": "unknown",
                    "retired_at": null,
                    "sale_price": "1.19"
                }
            },
            {
                "name": "Luke-Test-002",
                "container_id": null,
                "new_container_type": "mrv-medium",
                "destiny": {
                    "store": {
                        "where": "cold_4",
                        "shaking": false
                    }
                },
                "container_type": {
                    "id": "mrv-medium",
                    "name": "Medium MRV",
                    "well_count": 1,
                    "well_depth_mm": "81.5",
                    "well_volume_ul": "5000.0",
                    "capabilities": [
                        "agitate",
                        "cover",
                        "dispense",
                        "evaporate",
                        "image",
                        "incubate",
                        "liquid_handle",
                        "measure_mass",
                        "provision",
                        "spe",
                        "spin",
                        "uncover"
                    ],
                    "shortname": "mrv-medium",
                    "col_count": 1,
                    "is_tube": true,
                    "acceptable_lids": [
                        "screw-cap"
                    ],
                    "height_mm": "83.5",
                    "vendor": "unknown",
                    "catalog_number": "unknown",
                    "retired_at": null,
                    "sale_price": "1.19"
                }
            }
        ],
        "owner": {
            "id": "u1fkpc5skavpkm",
            "email": "luke.perrin@strateos.com",
            "name": "Luke Perrin",
            "profile_img_url": null
        },
        "project": {
            "id": "p1gvx9j9264ked",
            "name": "Luke-Test-002",
            "archived_at": null,
            "bsl": 1,
            "organization": {
                "id": "org1cytx5sk6tvss",
                "name": "L2S2 - Development",
                "subdomain": "l2s2dev"
            }
        },
        "successors": [],
        "instructions": [
            {
                "id": "i1gvxaxck7t44z",
                "sequence_no": 0,
                "operation": {
                    "op": "provision",
                    "resource_id": "rs1gstrsawf6hmy",
                    "measurement_mode": "mass",
                    "to": [
                        {
                            "well": "Luke-Test-002/0",
                            "mass": "12.212:milligram"
                        }
                    ]
                },
                "completed_at": null,
                "data_name": null,
                "started_at": null,
                "generated_containers": [],
                "warps": []
            },
            {
                "id": "i1gvxaxckcbvn5",
                "sequence_no": 1,
                "operation": {
                    "op": "provision",
                    "resource_id": "rs1gstrs8e6g4rp",
                    "measurement_mode": "volume",
                    "to": [
                        {
                            "well": "Luke-Test-002/0",
                            "volume": "0.021432:milliliter"
                        }
                    ]
                },
                "completed_at": null,
                "data_name": null,
                "started_at": null,
                "generated_containers": [],
                "warps": []
            },
            {
                "id": "i1gvxaxckfrg28",
                "sequence_no": 2,
                "operation": {
                    "op": "provision",
                    "resource_id": "rs1gstrseyaf3rw",
                    "measurement_mode": "volume",
                    "to": [
                        {
                            "well": "Luke-Test-002/0",
                            "volume": "0.6:milliliter"
                        }
                    ]
                },
                "completed_at": null,
                "data_name": null,
                "started_at": null,
                "generated_containers": [],
                "warps": []
            },
            {
                "id": "i1gvxaxckhzvaa",
                "sequence_no": 3,
                "operation": {
                    "op": "provision",
                    "resource_id": "rs1dj9bgfz3ks7r",
                    "measurement_mode": "volume",
                    "to": [
                        {
                            "well": "Luke-Test-002/0",
                            "volume": "1000:microliter"
                        }
                    ],
                    "informatics": [
                        {
                            "type": "attach_compounds",
                            "data": {
                                "wells": [
                                    "Luke-Test-002/0"
                                ],
                                "compounds": [
                                    {
                                        "format": "Daylight Canonical SMILES",
                                        "value": "CNC(=C1CC1)C1C=CC=C1"
                                    }
                                ]
                            }
                        }
                    ]
                },
                "completed_at": null,
                "data_name": null,
                "started_at": null,
                "generated_containers": [],
                "warps": []
            },
            {
                "id": "i1gvxaxcknefpd",
                "sequence_no": 4,
                "operation": {
                    "op": "microwave",
                    "object": "Luke-Test-002",
                    "duration": "10:minute",
                    "target_temperature": "70:celsius",
                    "pressure": "5:bar",
                    "agitate": {
                        "mode": "stir_bar",
                        "speed": "300:rpm"
                    },
                    "temperature_sensor_location": "chamber",
                    "pre_condition": false
                },
                "completed_at": null,
                "data_name": null,
                "started_at": null,
                "generated_containers": [],
                "warps": []
            },
            {
                "id": "i1gvxaxckru33g",
                "sequence_no": 5,
                "operation": {
                    "op": "provision",
                    "resource_id": "rs1dj945e3h695b",
                    "measurement_mode": "volume",
                    "to": [
                        {
                            "well": "Luke-Test-002-crude-LCMS-1/0",
                            "volume": "1:milliliter"
                        }
                    ]
                },
                "completed_at": null,
                "data_name": null,
                "started_at": null,
                "generated_containers": [],
                "warps": []
            },
            {
                "id": "i1gvxaxckv8nfk",
                "sequence_no": 6,
                "operation": {
                    "op": "liquid_handle",
                    "locations": [
                        {
                            "location": "Luke-Test-002/0",
                            "transports": [
                                {
                                    "volume": "-25:microliter"
                                }
                            ]
                        },
                        {
                            "location": "Luke-Test-002-crude-LCMS-1/0",
                            "transports": [
                                {
                                    "volume": "25:microliter"
                                }
                            ]
                        }
                    ],
                    "shape": {
                        "rows": 1,
                        "columns": 1,
                        "format": "SBS96"
                    },
                    "mode": "air_displacement"
                },
                "completed_at": null,
                "data_name": null,
                "started_at": null,
                "generated_containers": [],
                "warps": []
            },
            {
                "id": "i1gvxaxckzseyq",
                "sequence_no": 7,
                "operation": {
                    "op": "cover",
                    "object": "Luke-Test-002",
                    "lid": "screw-cap"
                },
                "completed_at": null,
                "data_name": null,
                "started_at": null,
                "generated_containers": [],
                "warps": []
            },
            {
                "id": "i1gvxaxcm42u8s",
                "sequence_no": 8,
                "operation": {
                    "op": "cover",
                    "object": "Luke-Test-002-crude-LCMS-1",
                    "lid": "screw-cap"
                },
                "completed_at": null,
                "data_name": null,
                "started_at": null,
                "generated_containers": [],
                "warps": []
            }
        ],
        "batch": {
            "id": "bat1gvxa6ndcemqb",
            "organization_id": "org1cytx5sk6tvss",
            "compound_link_id": "cmpl1fk274j27n7ve",
            "reaction_id": "5b90b624-e1b8-4944-b54d-cf0ed2d7d398",
            "purity": null,
            "post_purification_mass_yield_mg": null,
            "deleted_at": null,
            "created_at": "2022-04-04T15:15:02.418-07:00",
            "updated_at": "2022-04-04T15:15:02.418-07:00",
            "product_type": "FINAL_PRODUCT",
            "samples_created_at": null
        }
    }
]
```

New:
```
{
    "data": [
        {
            "id": "r1gvxaxcjwqh3r",
            "type": "runs",
            "links": {
                "self": "https://secure.strateos.com/api/runs/r1gvxaxcjwqh3r"
            },
            "attributes": {
                "title": "Synth_Luke-Test-002",
                "status": "canceled",
                "created_at": "2022-04-04T15:25:24.666-07:00",
                "completed_at": null
            }
        }
    ],
    "meta": {
        "record_count": 1
    },
    "links": {
        "first": "https://secure.strateos.com/api/runs?fields%5Bruns%5D=id%2Ctitle%2Cstatus%2Ccreated_at%2Ccompleted_at&filter%5Bproject_id%5D=p1gvx9j9264ked&page%5Blimit%5D=100&page%5Boffset%5D=0",
        "last": "https://secure.strateos.com/api/runs?fields%5Bruns%5D=id%2Ctitle%2Cstatus%2Ccreated_at%2Ccompleted_at&filter%5Bproject_id%5D=p1gvx9j9264ked&page%5Blimit%5D=100&page%5Boffset%5D=0"
    }
}
```

I'm not saying this is an issue necessarily, it's just grounds for a clear changelog message and minor version bump.

Reply:
Gotcha. It depends on what is the scope that we want to support. Do we want to support for users every single method available or only the ones available through the cli? I'll update the changelog to contain the limiting of the fields as well.